lib/block_view_stake.go

```diff
@@ -0,0 +1,2600 @@
+	if err != nil {
+		// We don't want to error if the key isn't found. Instead, return nil.
+		if err == badger.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "DBGetStakeEntry: problem retrieving StakeEntry: ")
+	}
+
+	// Decode StakeEntry from bytes.
+	rr := bytes.NewReader(stakeEntryBytes)
+	stakeEntry, err := DecodeDeSoEncoder(&StakeEntry{}, rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DBGetStakeEntry: problem decoding StakeEntry: ")
+	}
+	return stakeEntry, nil
+}
+
+func DBGetStakeEntriesForValidatorPKID(handle *badger.DB, snap *Snapshot, validatorPKID *PKID) ([]*StakeEntry, error) {
+	// Retrieve StakeEntries from db.
+	prefix := DBKeyForStakeByValidator(&StakeEntry{ValidatorPKID: validatorPKID})
+	_, valsFound, err := EnumerateKeysForPrefixWithLimitOffsetOrder(
+		handle, prefix, 0, nil, false, NewSet([]string{}),
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DBGetStakeEntriesForValidatorPKID: problem retrieving StakeEntries: ")
+	}
+
+	// Decode StakeEntries from bytes.
+	var stakeEntries []*StakeEntry
+	for _, stakeEntryBytes := range valsFound {
+		rr := bytes.NewReader(stakeEntryBytes)
+		stakeEntry, err := DecodeDeSoEncoder(&StakeEntry{}, rr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "DBGetStakeEntriesForValidatorPKID: problem decoding StakeEntry: ")
+		}
+		stakeEntries = append(stakeEntries, stakeEntry)
+	}
+	return stakeEntries, nil
+}
+
+func DBGetLockedStakeEntry(
+	handle *badger.DB,
+	snap *Snapshot,
+	validatorPKID *PKID,
+	stakerPKID *PKID,
+	lockedAtEpochNumber uint64,
+) (*LockedStakeEntry, error) {
+	var ret *LockedStakeEntry
+	err := handle.View(func(txn *badger.Txn) error {
+		var innerErr error
+		ret, innerErr = DBGetLockedStakeEntryWithTxn(
+			txn, snap, validatorPKID, stakerPKID, lockedAtEpochNumber,
+		)
+		return innerErr
+	})
+	return ret, err
+}
+
+func DBGetLockedStakeEntryWithTxn(
+	txn *badger.Txn,
+	snap *Snapshot,
+	validatorPKID *PKID,
+	stakerPKID *PKID,
+	lockedAtEpochNumber uint64,
+) (*LockedStakeEntry, error) {
+	// Retrieve LockedStakeEntry from db.
+	key := DBKeyForLockedStakeByValidatorByStakerByLockedAt(&LockedStakeEntry{
+		ValidatorPKID:       validatorPKID,
+		StakerPKID:          stakerPKID,
+		LockedAtEpochNumber: lockedAtEpochNumber,
+	})
+	lockedStakeEntryBytes, err := DBGetWithTxn(txn, snap, key)
+	if err != nil {
+		// We don't want to error if the key isn't found. Instead, return nil.
+		if err == badger.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(
+			err, "DBGetLockedStakeEntry: problem retrieving LockedStakeEntry: ",
+		)
+	}
+
+	// Decode LockedStakeEntry from bytes.
+	rr := bytes.NewReader(lockedStakeEntryBytes)
+	lockedStakeEntry, err := DecodeDeSoEncoder(&LockedStakeEntry{}, rr)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err, "DBGetLockedStakeEntry: problem decoding LockedStakeEntry: ",
+		)
+	}
+	return lockedStakeEntry, nil
+}
+
+func DBGetLockedStakeEntriesInRange(
```

Comment:
Would delete these if you decide to make _connectUnlock just unlock whatever the user is capable of unlocking.